miio/gateway/gateway.py

```diff
@@ -222,24 +222,34 @@ def get_devices_from_dict(self, device_dict):
         self._devices = {}
 
         # find the gateway
+        found_gateway = False
         for device in device_dict:
-            if device["mac"] == self.mac:
-                self._did = device["did"]
-                break
+            if device["did"] == str(self.device_id):
```

Comment:
Please refactor this into an internal `_get_device_by_did()` method which returns either None or the device object.
This way you can avoid deeper nesting inside this method, and make it much simpler by doing something like this:
```
gateway = self._get_device_by_did(self.device_id)`
if gateway is None:
    # log error
    return self._devices

if gateway["mac"] != self.mac:
    # log error about incorrect mac
    return self._devices

# here the rest of the code
```

Reply:
Done